### ISSUE TITLE
JupyterLab issues with Environment Modules (on remote server)

### ISSUE THREAD

**Issue by ThePrefixCyber** [jupyterlab/jupyterlab#5519]
I realize this could be a configuration issue for me, but I have invested considerable time into debugging and I think this is most likely either a bug or a missing feature.  Please tell me if I should be opening a "request for assistance" ticket instead.  

––––––

My school has a remote server with a bunch of preloaded software.  Users navigate and use software with [Environment Modules](https://uisapp2.iu.edu/confluence-prd/pages/viewpage.action?pageId=115540061).  The server is running CentOS 6.4, and my local system is Mac OS Sierra.  

Classic Jupyter Notebooks have no issues with this module system.  If I run:

`module load python/3.6.2`
`module load tensorflow`
`jupyter notebook`

Then open foo.ipynb in the navigator, and run a cell containing 'import tensorflow', tensorflow will import successfully.  

However, if I run:

`module load python/3.6.2`
`module load tensorflow`
`jupyter-lab --port=9000 --no-browser &`

(note: I have to do this forwarding step because launching JupyterLab in the server environment does not work at all. You can see instructions for this [here](https://www.blopig.com/blog/2018/03/running-jupyter-notebook-on-a-remote-server-via-ssh/) if you have issues getting it working.  I have confirmed that this step is not the issue, because the same thing can be done for classic Jupyter notebooks and there is no problem)

And then on my local:

`ssh -N -f -L 8888:localhost:9000 <user>@<server>`

JupyterLab runs cleanly from my local browser, however if I open a Jupyter notebook inside JupyterLab and run a cell containing `import tensorflow`, I get an ImportError.  

I don't want to dwell on the specifics for Tensorflow, but the point is that JupyterLab does not retain the configuration of the terminal that launched it, and (as far as I can tell) there is no way to access the environment that JupyterLab python code is running in.  In my case, I have tried modifying sys.path and all of the related environment variables from inside my Python script (LD_LIBRARY_PATH and several specific environment variables relating to Environment Modules), but this doesn't truly modify those variables in the environment, it simply overwrites the values of sys.path etc. within Python.  I can get Python to locate the tensorflow binary file, but then I get a number of other errors where it can't find other dependencies, such as the GCC compiler.   

Basically, there are likely a bunch of configuration scripts running on the server which do not launch when a Python kernel is launched in JupyterLab, and there doesn't seem to be a way to apply all of these to a kernel once it has been launched.  A workaround for my specific case would be nice, but in general it seems that this is a limitation of JupyterLab compared to classic Jupyter Notebooks.  Let me know if this is something that can be fixed, or if it's just a lack of understanding on my part. 

Thank you!

**Comment by blink1073** [comment]
Hi @ThePrefixCyber, if you use `jupyter lab` (without the dash, similar to how you launched `jupyter notebook`), do you pick up the correct import?

**Comment by ThePrefixCyber** [comment]
I realized that Jupyter Lab runs a new internal shell for the Python kernel, while classic Jupyter seems to run in the shell that launched it.  Jupyter Lab does call my user-specific .bashrc script on launch however, so even though I have not figured out how to configure the shell environment on-the-fly in Jupyter Lab, I can work around my issue by configuring everything in .bashrc ahead of time before launching.  This is good enough for my purposes.  

Please let me know if I am misunderstanding how Jupyter Lab works, and if there is a way to access the shell my code is running inside.  Feel free to close this issue if, in your judgment, this workaround is my best option.

**Comment by blink1073** [comment]
Do `which jupyter` and `which jupyter-lab` point to different folders?

**Comment by ThePrefixCyber** [comment]
Yes it seems that while I have installed Jupyter, Jupyter-Lab may be using a version of Jupyter which was already on the system.  If there are multiple Jupyter's installed on the server, is there a way to change the Jupyter which Jupyter-Lab uses, or at least specify during setup?

Also, responding to your above question, calling `jupyter-lab` vs `jupyter` lab does not seem to make any difference.

**Comment by blink1073** [comment]
I can't diagnose more without more information:

- which jupyter
- which jupyter-notebook
- which jupyter-lab

**Comment by ThePrefixCyber** [comment]
which jupyter
/share/pkg/python/3.6.2/install/bin/jupyter
(note this is not the python version that runs the code when I actually use JupyterLab)

which jupyter-notebook
/share/pkg/python/3.6.2/install/bin/jupyter-notebook

which jupyter-lab
~/.local/bin/jupyter-lab

I installed my own Jupyter in .local/bin/ but it doesn't seem to use it.

**Comment by blink1073** [comment]
There are two options:  

- remove `/share/pkg/python/3.6.2/install` from your PATH environment variable
- Install `jupyterlab` in `/share/pkg/python/3.6.2`

**Comment by blink1073** [comment]
Ah, actually I think you want to use a third option: prepend `~/.local/bin` to PATH.

**Comment by jasongrout** [comment]
Closing as answered - feel free to follow up here with comments if you'd like, though.
